### PR TITLE
[BEAM-1205] Auto detect and set "enableAbandonedNodeEnforcement" in TestPipeline

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @tgroh , @kennknowles

### CODE REVIEW

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineEnforcementsTest.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    }
+
+    @Category(RunnableOnService.class)
+    @Test
+    public void testMissingRunAutoAdd_runnableOnService_expectNoException()
+        throws Exception {
+      pipeline.enableAutoRunIfMissing(true);
+      addTransform(pCollection(pipeline));
+
+      // have the the pipeline.run() auto-added
```
Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineEnforcementsTest.java]: I suppose one "the" would suffice here...

**Comment by kennknowles** [comment]
To be honest, even though we violate this in a few places, I strongly prefer to have tests always named after a real class, for discoverability.

**Comment by staslev** [comment]
Makes sense, I'll make the proper arrangements :)

Let me know if you spot any other issues.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java [hunk 2]
```diff
@@ -239,8 +246,53 @@ private TestPipeline(
...
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation annotation) {
+          return annotation.annotationType() != null
+              && annotation.annotationType().equals(Category.class);
+        }
+      };
+
+      private final Predicate<Annotation> categoryOfNeedsRunner = new Predicate<Annotation>() {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: Can this be a constant? `HAS_NEEDS_RUNNER_CATEGORY`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java [hunk 3]
```diff
@@ -239,8 +246,53 @@ private TestPipeline(
   public Statement apply(final Statement statement, final Description description) {
     return new Statement() {
 
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: This can be a constant `IS_CATEGORY_ANNOTATION`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java [hunk 4]
```diff
@@ -239,8 +246,53 @@ private TestPipeline(
...
+          final boolean annotatedWithNeedsRunner =
+              FluentIterable
+                  .from(description.getAnnotations())
+                  .filter(hasCategoryAnnotation)
+                  .anyMatch(categoryOfNeedsRunner);
+
+          final boolean crashingRunner =
+              CrashingRunner.class.isAssignableFrom(getOptions().getRunner());
+
+          enableAbandonedNodeEnforcement(annotatedWithNeedsRunner || !crashingRunner);
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: If a test is annotated with `NeedsRunner` and has a `CrashingRunner`, there's a configuration problem in whatever is running the tests. It might be worth having a `checkState` here to validate that.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java [hunk 5]
```diff
@@ -239,8 +246,53 @@ private TestPipeline(
...
+                    @Override
+                    public boolean apply(final Class<?> aClass) {
+                      return NeedsRunner.class.isAssignableFrom(aClass);
+                    }
+                  });
+        }
+      };
+
+      private void setDeducedEnforcementLevel() {
+        if (enforcement instanceof PipelineAbandonedNodeEnforcement) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: What if the user explicitly calls `enableAbandonedNodeEnforcement`? I would have enforcement start as `null` to ensure we don't clobber any explicit calls, so we infer the enforcement iff it hasn't been explicitly set.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 3 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: That's a good point.

Since `hasCategoryAnnotation` is used only in the enclosing (anonymous) `Statement` instance I thought it would be nice to place it in the most inner scope, i.e. the anonymous `Statement` class, in which case it cannot be static due to it being anonymous.

Do you believe making it a constant outweighs the scoping consideration?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 2 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: Please see the comment regarding IS_CATEGORY_ANNOTATION.

If we decide to go constant over scoping I'll apply the change here as well.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 5 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: That's a really good point. 

First, if a user changes the enforcement level inside the test code, it will take effect since we call ` statement.evaluate()` (i.e. the user code) **after** `setDeducedEnforcementLevel` is called.
```java
@Override
      public void evaluate() throws Throwable {
        setDeducedEnforcementLevel(); // set the enforcement level we think should be set
        statement.evaluate(); //user code is run here, so any changes will overwrite our pervious setting for the enforcement level
        enforcement.afterTestCompletion();
      }
```

If the user changes the enforcement level at the rule level, i.e. outside of the test, e.g.  `@Rule TestPipeline = TestPipeline.create().enableAbandonedNodeEnforcement(...)` it is handled inside `setDeducedEnforcementLevel()`
```java
if (enforcement instanceof PipelineAbandonedNodeEnforcement) { ... }
 // if enforcement instanceof PipelineRunEnforcement at this point, it must have been
// set so by the user, so don't second guess it.
```

Basically this should cover both cases of users change enforcement level inside and outside their tests.

Do you believe we have other use-cases that need handling?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 4 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: Sounds good, I'll add a check here 👍

**Comment by staslev** [comment]
@tgroh , thanks for taking a look.

I've addressed your comments, let me know how it looks.

**Comment by staslev** [comment]
@tgroh , @kennknowles I was wondering if you had any further comments before we can move forward with this?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: I have a minor preference, but not enough to demand it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: For the first portion, that's a good point - apparently when I think about a test method I think of it like how a Pipeline is run (e.g. build then execute, where execute == `p.run()`)

For the second portion, I don't quite understand what you mean - if the user sets `enableAbandonedNodeEnforcement` before the call to evaluate (happens as part of `@Rule TestP...setAbandonedNodeEnforcement...`, yes?), this will still do the inference as though that method was not called right?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 6]
```diff
@@ -52,280 +59,491 @@
...
+      @Rule
+      public final transient TestPipeline pipeline = TestPipeline.fromOptions(dummyRunner());
 
-        @Override
-        public State waitUntilFinish(final Duration duration) {
-          return null;
-        }
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNormalFlow_runnableOnService_expectNoException() throws Exception {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Our normal test naming strategy is without underscores - I'm willing to rename these before merge, or you can do so.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 7]
```diff
@@ -52,280 +59,491 @@
...
+                                   Request.method(testClass, description.getMethodName()),
+                                   description.getTestClass().getSimpleName() + "#"
+                                       + description.getMethodName()
+                               };
+                             }
+                           })
+                           .toList();
+    }
+
+    private TestRule withPossibleExceptions(final String testName) {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: This worries me.

I'm ok with having the tests run with `Parameterized`, but if we do so we should still be able to determine which case we're in and set the expected exception per-test within the test based on the value of the parameters, rather than inspecting the name of the test. For example, `testConflict_runnableOnServiceAnnotation_expectIllegalStateException` doesn't demonstrate by reading the test code directly that it expects an exception.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 5 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: Not exactly. `setDeducedEnforcementLevel()` "knows" that the default enforcement is `PipelineAbandonedNodeEnforcement`, so in case it detects it's been called with the enforcement not being the default it will not run the inference logic and leave things as they are, or better yet, as the user has set them up.

```java 
if (enforcement instanceof PipelineAbandonedNodeEnforcement /* which is the default */ ) {
 // do inference 
} else /* enforcement is not the default, so the user must have changed it */ { 
  // bail out, do not execute inference logic
}
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java (hunk 6 above)
Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Sorry I did not provide more context here.
`NonCrachingRunner` and `WithCrachingRunner` are kind of meta-tests here. Since I need to test the `TestPipeline` rule in various scenarios, including with and without `@NeedsRunner/RunnableOnService` I use `JUnitCore` to run these tests myself (since otherwise they use the `pom` definitions). In a way, I create a test bed for tests since I would like to test things by writing "user scenarios".

I use this funny looking naming to make the test-bed generic and be able to specify what's the expected behaviour, see also `runTest`.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java (hunk 7 above)
Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: This is also a part of the test-bed mechanism I described earlier, hopefully the previous comment provides some more context, let me know if it makes sense.

**Comment by staslev** [comment]
@tgroh thanks for taking a look!

Hopefully I was able to provide more context on some of the design decisions behind this PR.
Please feel free to let me know if I can provide more info.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 5 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java]: I don't think I understand your first comment - if a user writes `@Rule TestPipeline = TestPipeline.create().enableAbandonedNodeEnforcement(...)` it will still be automatically downgraded if `crashingRunner && !annotatedWithNeedsRunner`. This probably means that a user forgot `NeedsRunner` or shouldn't be using `TestPipeline`, so should fail instead.

Regardless of how we isolate the use cases, I think the whole thing will probably be more readable if it is just very explicit whether it has been set versus inferred. I tend to prefer `Optional` to `@Nullable` for safety, readability and composability.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 8]
```diff
@@ -52,280 +59,491 @@
...
+        };
+      }
     }
-  }
 
-  private static class DummyRunner extends PipelineRunner<PipelineResult> {
+    @SuppressWarnings("UnusedReturnValue")
+    private static PCollection<String> addTransform(final PCollection<String> pCollection) {
+      return pCollection.apply("Map2",
+                               MapElements.via(new SimpleFunction<String, String>() {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: FWIW we go with `google-java-format`. We can autoformat when merging, but it is most readable to just get that set up in your IDE or via command line.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 9]
```diff
@@ -52,280 +59,491 @@
...
+    private TestRule withPossibleExceptions(final String testName) {
+      final String expect = testName.substring(testName.indexOf("_expect"));
+      if (expect.contains(TestPipeline.PipelineRunMissingException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.PipelineRunMissingException.class);
+        return exception;
+      } else if (expect.contains(TestPipeline.AbandonedNodeException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        if (testName.contains(P_TRANSFORM)) {
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: The expectations set up here can be utility methods, but their calls should be inlined to the test methods themselves. Drive the test with code, not its name. It actually makes it very troublesome to find out what a test is really doing if one needs to go look in these shared conditionals based on string munging.

Instead of:

```java
@Test
@RunnableOnService
public void testConflict_runnableOnServiceAnnotation_expectIllegalStateException()
    throws Exception {
  addTransform(pCollection(pipeline));
  // @RunnableOnService annotation with a CrashingRunner indicates a configuration issue
}		
```

this:

```java
/**
 * A `RunnableOnService` category with a `CrashingRunner` indicates a configuration issue.
 */
@Test
public void testCrashingRunnerRunnableOnService() throws Exception {
  thrown.expect(IllegalStateException.class);
  thrown.expectMessage("some keywords, but not too many, to make the message has pertinent info");
  addTransform(pCollection(pipeline));
}
```

Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: @kennknowles thanks for taking a look. I had intentionally jumped through hoops here in order to make these test look exactly like the user would write them, without the stuff WE need to test them (e.g., `thrown.expect` and the likes). This is why I chose to make the expected exceptions external to the tests, users would not expect TestPipeline exceptions to be thrown, instead I used the naming to figure them out outside the test code (i.e. on the meta-layer).

Do you believe we should move away from these tests representing users scenarios 1:1 in favour of having the expected exceptions inside the tests?

**Comment by staslev** [comment]
@kennknowles , I have revised according to (most of) your comments.

Essentially this leaves with the discussion about "having 1:1 mapping between user-scenarios and tests vs. explicit exception handling inside the tests" (more details in the comments above).

**Comment by staslev** [comment]
@kennknowles , @tgroh  I caved in to your arguments around the enforcement tests :)

Let me know if there's more or if I have missed anything.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 10]
```diff
@@ -52,280 +59,496 @@
...
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+        pipeline.run().waitUntilFinish();
+
+        // dangling PAssert
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNoTestPipelineUsedRunnableOnService() { }
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: This is a really strange test - It's at least worth commenting what it's testing more explicitly (e.g. 
"The `{@code TestPiplieline}` is initialized as a field in the test class, but is not used in this test; with the default enforcement active, this test succeeds".

The same is true of the test immediately following, and any other empty test.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 11]
```diff
@@ -52,280 +59,496 @@
...
-  @Test
-  public void testAutoAddMissingRunFlow() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    // missing pipeline#run, but have it auto-added.
-    pipeline.enableAutoRunIfMissing(true);
-  }
+    }
+
+
+    private static List<Object[]> extractTests(final Class<?> testClass) {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Can this be removed?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/Annotations.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.common.collect.FluentIterable;
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import javax.annotation.Nonnull;
+import org.junit.experimental.categories.Category;
+
+/**
+ * A utility class for querying annotations.
+ */
+public class Annotations {
```
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/Annotations.java]: 👍

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java (hunk 10 above)
Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: I'll add some more javadoc.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java (hunk 11 above)
Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: The blank line before the method? Or the method itself?

Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Both; my understanding was that we would move all of the test to the style used in the rest of the project.

Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: I thought we were talking about the naming convention only,

I have resorted to managing the test execution myself because otherwise tests annotated with `NeedsRunner` and `RunnableOnService` are not run (due to exclusion in `pom.xmll`). 

In order to test `TestPipeline` under various scenarios, including tests annotated with `NeedsRunner` and `RunnableOnService` (but regardless of the runner) I use the above mentioned method to extract tests and run them manually using `JUnitCore` (see `TestPipelineTest.TestPipelineEnforcementsTest#runTest`).

If there's a cleaner way to achieve this, please do share :)

Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: The tests should be run in the context of the `runners/direct` module, at minimum, so the same signal should be obtained. Failing that, adding an additional category and explicitly including it during the execution of the core SDK should also work (even if it adds an additional Surefire execution to the core SDK)

Comment (staslev) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: Why do you believe they should run in the context of `runners/direct`?
These tests are quite agnostic of actual runners, which is why they use internal dummy runners (i.e. `DummyRunner` and `CrashingRunner`). Since this feature is not affected by the various beam runners, I thought the `sdk` module would be an ideal fit.

Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: It's less that they "should" be run in the context of "runners/direct" than that they will (which means we don't lose coverage); I'd rather they run in the wrong module (but always run) than have to maintain custom test execution code. Plus, as the alternative these could be in some new category ("`TestInfrastructure`" or something) and explicitly included the same way the runners scan for `NeedsRunner`/`RunnableOnService` tests

**Comment by staslev** [comment]
@tgroh made some changes following our discussion.

PTAL.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java [hunk 13]
```diff
@@ -24,308 +25,375 @@
...
+      @Category(RunnableOnService.class)
+      @Test
+      public void testDanglingPTransformRunnableOnService() throws Exception {
 
-        @Override
-        public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
-            throws AggregatorRetrievalException {
-          return null;
-        }
+        exception.expect(TestPipeline.AbandonedNodeException.class);
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java]: I would move the expectations to occur immediately before the throwing statement.

**Comment by staslev** [comment]
Rebased and moved expectations to occur before throwing statements.

Waiting for tests to pass, then merging.

**Comment by staslev** [comment]
Reformatted the changed code using the `google-code-format` [plugin for IntelliJ](https://plugins.jetbrains.com/idea/plugin/8527-google-java-format).
